include/GooseEYE/Ensemble_L.hpp

```diff
@@ -0,0 +1,119 @@
+/* =================================================================================================
+
+(c - GPLv3) T.W.J. de Geus (Tom) | tom@geus.me | www.geus.me | github.com/tdegeus/GooseEYE
+
+TODO:
+- Normalization
+- Account for mask
+- Comparison sub-matrix is now copied, passing a view to the roi_L function would be faster
+
+================================================================================================= */
+
+#ifndef GOOSEEYE_ENSEMBLE_L_HPP
+#define GOOSEEYE_ENSEMBLE_L_HPP
+
+#include "GooseEYE.h"
+
+namespace GooseEYE {
+
+// -------------------------------------------------------------------------------------------------
+
+template <class T>
+inline void Ensemble::L(
+  const xt::xarray<T>& f,
+  const xt::xarray<int>& fmask
+)
+{
+
+  // lock statistics
+  m_stat = Type::L;
+
+  // padding default not periodic: mask padded items
+  xt::pad_mode pad_mode = xt::pad_mode::constant;
+  int mask_value = 1;
+
+  // padding optionally periodic: unmask padded items
+  if (m_periodic) {
+    pad_mode = xt::pad_mode::periodic;
+    mask_value = 0;
+  }
+
+  // apply padding
+  xt::xarray<T>   F = xt::pad(f, m_pad, pad_mode);
+  xt::xarray<int> Fmask = xt::pad(fmask, m_pad, xt::pad_mode::constant, mask_value);
+
+  // make matrices virtually 3-d matrices
+  std::vector<size_t> shape = detail::shape_as_dim(MAX_DIM, F, 1);
+  F.reshape(shape);
+  Fmask.reshape(shape);
+
+  // local output and normalisation
+  xt::xarray<T> first = xt::zeros<T>(m_Shape);
+  xt::xarray<T> norm = xt::zeros<T>(m_Shape);
+
+  // sub-matrix and path-matrix
+  xt::xarray<T> Fi = xt::empty<T>(m_Shape);
+  xt::xarray<int> Li = xt::empty<T>(m_Shape);
+
+  // roi stamp
+  xt::xarray<int> stamp = stampPoints( 3 );
+  size_t nstamp = stamp.shape()[0];
+
+  // Position vectors
+  std::array<int,3> x0, x1;
+
+  x0[0] = (int) m_Shape[0] / 2;
+  x0[1] = (int) m_Shape[1] / 2;
+  x0[2] = (int) m_Shape[2] / 2;
```

Comment:
This implementation is based on an odd-sized roi. x0 is the mid-point of the roi obtained by integer division of the shape by 2. Should the software be able to handle an even-sized roi (which would be tricky since there is no center pixel), or should it just throw an error in that case.

Reply:
In my memory...: In the new implementation functions should be general, allowing both even and odd shaped ROIs. Though I have no problem restricting specific statistics to odd-only, with the proper check.